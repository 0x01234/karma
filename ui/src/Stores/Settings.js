import { localStored } from "mobx-stored";

import { action } from "mobx";

const defaultSavedFilters = {
  filters: [],
  present: false
};

class Settings {
  savedFilters = localStored("savedFilters", defaultSavedFilters, {
    delay: 100
  });

  saveFilters = action(newFilters => {
    this.savedFilters.filters = newFilters;
    this.savedFilters.present = true;
  });

  clearSavedFilters = action(() => {
    this.savedFilters.filters = [];
    this.savedFilters.present = false;
  });
}

export { Settings };
