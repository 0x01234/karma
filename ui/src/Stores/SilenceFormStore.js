import { observable, action, computed } from "mobx";

import uniqueId from "lodash.uniqueid";

import moment from "moment";

const NewEmptyMatcher = id => {
  return {
    id: uniqueId(),
    name: "",
    values: [],
    suggestions: {
      names: [],
      values: []
    },
    isRegex: false
  };
};

const ValueToObject = value => ({ label: value, value: value });

class SilenceFormStore {
  // this is used to store modal visibility toggle
  toggle = observable(
    {
      visible: false,
      toggle() {
        this.visible = !this.visible;
      },
      hide() {
        this.visible = false;
      },
      show() {
        this.visible = true;
      }
    },
    { toggle: action.bound, hide: action.bound, show: action.bound }
  );

  // form data is stored here, it's global (rather than attached to the form)
  // so it can be manipulated from other parts of the code
  // example: when user clicks a silence button on alert we should populate
  // this form from that alert so user can easily silence that alert
  data = observable(
    {
      inProgress: false,
      alertmanagers: [],
      matchers: [],
      startsAt: moment(),
      endsAt: moment().add(1, "hour"),
      comment: "",
      author: "",

      resetProgress() {
        this.inProgress = false;
      },

      // append a new empty matcher to the list
      addEmptyMatcher() {
        let m = NewEmptyMatcher();
        this.matchers.push(m);
      },

      deleteMatcher(id) {
        // only delete matchers if we have more than 1
        if (this.matchers.length > 1) {
          this.matchers = this.matchers.filter(m => m.id !== id);
        }
      },

      fillMatchersFromGroup(group) {
        let matchers = [];

        // add matchers for all shared labels in this group
        for (const [key, value] of Object.entries(
          Object.assign({}, group.labels, group.shared.labels)
        )) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [ValueToObject(value)],
            suggestions: {
              names: [],
              values: []
            },
            isRegex: false
          });
        }

        // add matchers for all unique labels in this group
        let labels = {};
        for (const alert of group.alerts) {
          for (const [key, value] of Object.entries(alert.labels)) {
            if (!labels[key]) {
              labels[key] = new Set();
            }
            labels[key].add(value);
          }
        }
        for (const [key, values] of Object.entries(labels)) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [...values].sort().map(value => ValueToObject(value)),
            suggestions: {
              names: [],
              values: []
            },
            isRegex: values.size > 1
          });
        }

        this.matchers = matchers;
      },

      verifyStarEnd() {
        if (this.endsAt.isSameOrBefore(this.startsAt)) {
          this.endsAt = moment(this.startsAt).add(1, "minutes");
        }
      },
      incStart(minutes) {
        this.startsAt = moment(this.startsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decStart(minutes) {
        this.startsAt = moment(this.startsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incEnd(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decEnd(minutes) {
        this.endsAt = moment(this.endsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incDuration(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
      },
      decDuration(minutes) {
        const newEndsAt = moment(this.endsAt).subtract(minutes, "minutes");
        if (newEndsAt.isAfter(this.startsAt)) {
          this.endsAt = newEndsAt;
        }
      },

      get toAlertmanagerPayload() {
        const payload = {
          matchers: this.matchers.map(m => ({
            name: m.name,
            value:
              m.values.length > 1
                ? `(${m.values.map(v => v.value).join("|")})`
                : m.values.length === 1
                  ? m.values[0].value
                  : "",
            isRegex: m.isRegex
          })),
          startsAt: this.startsAt
            .second(0)
            .millisecond(0)
            .toISOString(),
          endsAt: this.endsAt
            .second(59)
            .millisecond(0)
            .toISOString(),
          createdBy: this.author,
          comment: this.comment
        };
        return payload;
      },
      get toDuration() {
        const data = {
          days: this.endsAt.diff(this.startsAt, "days"),
          hours: this.endsAt.diff(this.startsAt, "hours") % 24,
          minutes: this.endsAt.diff(this.startsAt, "minutes") % 60
        };
        return data;
      }
    },
    {
      resetProgress: action.bound,
      addEmptyMatcher: action.bound,
      deleteMatcher: action.bound,
      fillMatchersFromGroup: action.bound,
      verifyStarEnd: action.bound,
      incStart: action.bound,
      decStart: action.bound,
      incEnd: action.bound,
      decEnd: action.bound,
      incDuration: action.bound,
      decDuration: action.bound,
      toAlertmanagerPayload: computed,
      toDuration: computed
    },
    { name: "Silence form store" }
  );
}

export { SilenceFormStore };
