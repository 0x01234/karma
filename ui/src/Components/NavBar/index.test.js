import React from "react";

import { shallow, mount } from "enzyme";

import moment from "moment";

import { AlertStore, NewUnappliedFilter } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { NavBar, NavbarOnResize } from ".";

let alertStore;
let settingsStore;
let silenceFormStore;

beforeEach(() => {
  alertStore = new AlertStore([]);
  settingsStore = new Settings();
  silenceFormStore = new SilenceFormStore();
  // fix startsAt & endsAt dates so they don't change between tests
  silenceFormStore.data.startsAt = moment([2018, 1, 30, 10, 25, 50]).utc();
  silenceFormStore.data.endsAt = moment([2018, 1, 30, 11, 25, 50]).utc();
});

const RenderNavbar = () => {
  return shallow(
    <NavBar
      alertStore={alertStore}
      settingsStore={settingsStore}
      silenceFormStore={silenceFormStore}
    />
  ).find(".container");
};

const MountedNavbar = () => {
  return mount(
    <NavBar
      alertStore={alertStore}
      settingsStore={settingsStore}
      silenceFormStore={silenceFormStore}
    />
  );
};

const ValidateNavClass = (totalFilters, expectedClass) => {
  for (let i = 0; i < totalFilters; i++) {
    alertStore.filters.values.push(NewUnappliedFilter(`foo=${i}`));
  }
  const tree = RenderNavbar();
  const nav = tree.find(".navbar-nav");
  expect(nav.props().className.split(" ")).toContain(expectedClass);
};

describe("<NavBar />", () => {
  it("matches snapshot with 0 alerts", () => {
    const tree = RenderNavbar();
    expect(tree.text()).toMatchSnapshot();
  });

  it("matches snapshot with 5 alerts", () => {
    alertStore.info.totalAlerts = 5;
    const tree = RenderNavbar();
    expect(tree.text()).toMatchSnapshot();
  });

  it("navbar-brand shows 15 alerts with totalAlerts=15", () => {
    alertStore.info.totalAlerts = 15;
    const tree = RenderNavbar();
    const brand = tree.find(".navbar-brand");
    expect(brand.text()).toBe("15<FetchIndicator />");
  });

  it("navbar-nav includes 'flex-row' class with 0 filters", () => {
    ValidateNavClass(0, "flex-row");
  });

  it("navbar-nav includes 'flex-row' class with 1 filter", () => {
    ValidateNavClass(1, "flex-row");
  });

  it("navbar-nav includes 'flex-column' class with 2 filters", () => {
    ValidateNavClass(2, "flex-column");
  });

  it("navbar-nav includes 'flex-column' class with 3 filters", () => {
    ValidateNavClass(3, "flex-column");
  });
});

describe("NavbarOnResize()", () => {
  it("body 'padding-top' style is updated after calling NavbarOnResize()", () => {
    NavbarOnResize(0, 10);
    expect(
      window
        .getComputedStyle(document.body, null)
        .getPropertyValue("padding-top")
    ).toBe("14px");

    NavbarOnResize(0, 36);
    expect(
      window
        .getComputedStyle(document.body, null)
        .getPropertyValue("padding-top")
    ).toBe("40px");
  });
});

describe("<IdleTimer />", () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  it("hides navbar after 4 minutes", () => {
    const tree = MountedNavbar();
    expect(tree.find(".navbar")).toHaveLength(1);

    jest.runTimersToTime(1000 * 60 * 4);
    tree.update();
    expect(tree.find(".navbar")).toHaveLength(0);
  });

  it("hidden navbar shows up again after activity", () => {
    const tree = MountedNavbar();
    const instance = tree.instance();

    instance.activityStatus.idle = true;
    jest.runOnlyPendingTimers();
    tree.update();
    expect(tree.find(".navbar")).toHaveLength(0);

    instance.activityStatus.setActive();
    jest.runOnlyPendingTimers();
    tree.update();
    expect(tree.find(".navbar")).toHaveLength(1);
  });

  it("body padding-top is 4px when navbar is hidden", () => {
    const tree = MountedNavbar();
    const instance = tree.instance();

    instance.activityStatus.setIdle();
    jest.runOnlyPendingTimers();
    tree.update();
    expect(
      window
        .getComputedStyle(document.body, null)
        .getPropertyValue("padding-top")
    ).toBe("4px");
  });
});
