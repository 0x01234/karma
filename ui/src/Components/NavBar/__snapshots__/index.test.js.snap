// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar /> matches snapshot with 0 alerts 1`] = `
"
<div class=\\"container visible\\">
  <nav class=\\"navbar fixed-top navbar-expand navbar-dark p-1 bg-primary-transparent d-inline-block\\">
    <div style=\\"position:absolute;width:0;height:0;visibility:hidden;display:none\\">
    </div>
    <span class=\\"navbar-brand my-0 mx-2 h1 d-none d-sm-block float-left\\">
      0
      <svg aria-hidden=\\"true\\"
           data-prefix=\\"fas\\"
           data-icon=\\"circle-notch\\"
           class=\\"svg-inline--fa fa-circle-notch fa-w-16 fa-lg mx-1 text-muted\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 512 512\\"
           style=\\"opacity:0\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z\\"
        >
        </path>
      </svg>
    </span>
    <ul class=\\"navbar-nav float-right d-flex flex-row\\">
      <li class=\\"nav-item\\">
        <span class=\\"nav-link cursor-pointer\\">
          <svg aria-hidden=\\"true\\"
               data-prefix=\\"fas\\"
               data-icon=\\"bell-slash\\"
               class=\\"svg-inline--fa fa-bell-slash fa-w-20 \\"
               role=\\"img\\"
               xmlns=\\"http://www.w3.org/2000/svg\\"
               viewbox=\\"0 0 640 512\\"
          >
            <path fill=\\"currentColor\\"
                  d=\\"M633.82 458.1l-90.62-70.05c.19-1.38.8-2.66.8-4.06.05-7.55-2.61-15.27-8.61-21.71-19.32-20.76-55.47-51.99-55.47-154.29 0-77.7-54.48-139.9-127.94-155.16V32c0-17.67-14.32-32-31.98-32s-31.98 14.33-31.98 32v20.84c-40.33 8.38-74.66 31.07-97.59 62.57L45.47 3.37C38.49-2.05 28.43-.8 23.01 6.18L3.37 31.45C-2.05 38.42-.8 48.47 6.18 53.9l588.35 454.73c6.98 5.43 17.03 4.17 22.46-2.81l19.64-25.27c5.42-6.97 4.17-17.02-2.81-22.45zM157.23 251.54c-8.61 67.96-36.41 93.33-52.62 110.75-6 6.45-8.66 14.16-8.61 21.71.11 16.4 12.98 32 32.1 32h241.92L157.23 251.54zM320 512c35.32 0 63.97-28.65 63.97-64H256.03c0 35.35 28.65 64 63.97 64z\\"
            >
            </path>
          </svg>
        </span>
      </li>
      <li class=\\"nav-item\\">
        <span class=\\"nav-link cursor-pointer\\">
          <svg aria-hidden=\\"true\\"
               data-prefix=\\"fas\\"
               data-icon=\\"cog\\"
               class=\\"svg-inline--fa fa-cog fa-w-16 \\"
               role=\\"img\\"
               xmlns=\\"http://www.w3.org/2000/svg\\"
               viewbox=\\"0 0 512 512\\"
          >
            <path fill=\\"currentColor\\"
                  d=\\"M444.788 291.1l42.616 24.599c4.867 2.809 7.126 8.618 5.459 13.985-11.07 35.642-29.97 67.842-54.689 94.586a12.016 12.016 0 0 1-14.832 2.254l-42.584-24.595a191.577 191.577 0 0 1-60.759 35.13v49.182a12.01 12.01 0 0 1-9.377 11.718c-34.956 7.85-72.499 8.256-109.219.007-5.49-1.233-9.403-6.096-9.403-11.723v-49.184a191.555 191.555 0 0 1-60.759-35.13l-42.584 24.595a12.016 12.016 0 0 1-14.832-2.254c-24.718-26.744-43.619-58.944-54.689-94.586-1.667-5.366.592-11.175 5.459-13.985L67.212 291.1a193.48 193.48 0 0 1 0-70.199l-42.616-24.599c-4.867-2.809-7.126-8.618-5.459-13.985 11.07-35.642 29.97-67.842 54.689-94.586a12.016 12.016 0 0 1 14.832-2.254l42.584 24.595a191.577 191.577 0 0 1 60.759-35.13V25.759a12.01 12.01 0 0 1 9.377-11.718c34.956-7.85 72.499-8.256 109.219-.007 5.49 1.233 9.403 6.096 9.403 11.723v49.184a191.555 191.555 0 0 1 60.759 35.13l42.584-24.595a12.016 12.016 0 0 1 14.832 2.254c24.718 26.744 43.619 58.944 54.689 94.586 1.667 5.366-.592 11.175-5.459 13.985L444.788 220.9a193.485 193.485 0 0 1 0 70.2zM336 256c0-44.112-35.888-80-80-80s-80 35.888-80 80 35.888 80 80 80 80-35.888 80-80z\\"
            >
            </path>
          </svg>
        </span>
      </li>
    </ul>
    <form class=\\"form-inline mw-100\\"
          data-filters
    >
      <div role=\\"combobox\\"
           aria-haspopup=\\"listbox\\"
           aria-owns=\\"react-autowhatever-1\\"
           aria-expanded=\\"false\\"
           class=\\"autosuggest d-inline-block w-100\\"
      >
        <div class=\\"input-group mr-2\\">
          <div class=\\"input-group-prepend\\">
            <span class=\\"input-group-text px-2\\">
              <svg aria-hidden=\\"true\\"
                   data-prefix=\\"fas\\"
                   data-icon=\\"search\\"
                   class=\\"svg-inline--fa fa-search fa-w-16 \\"
                   role=\\"img\\"
                   xmlns=\\"http://www.w3.org/2000/svg\\"
                   viewbox=\\"0 0 512 512\\"
              >
                <path fill=\\"currentColor\\"
                      d=\\"M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z\\"
                >
                </path>
              </svg>
            </span>
          </div>
          <div class=\\"form-control p-1 components-filterinput\\">
            <input type=\\"text\\"
                   class=\\"components-filterinput-wrapper\\"
                   placeholder
                   value
                   autocomplete=\\"off\\"
                   aria-autocomplete=\\"list\\"
                   aria-controls=\\"react-autowhatever-1\\"
            >
          </div>
          <div class=\\"input-group-append\\">
            <button class=\\"input-group-text rounded-right cursor-pointer components-navbar-history px-2\\"
                    type=\\"button\\"
                    data-toggle=\\"dropdown\\"
                    aria-haspopup=\\"true\\"
                    aria-expanded=\\"true\\"
            >
              <svg aria-hidden=\\"true\\"
                   data-prefix=\\"fas\\"
                   data-icon=\\"caret-down\\"
                   class=\\"svg-inline--fa fa-caret-down fa-w-10 \\"
                   role=\\"img\\"
                   xmlns=\\"http://www.w3.org/2000/svg\\"
                   viewbox=\\"0 0 320 512\\"
              >
                <path fill=\\"currentColor\\"
                      d=\\"M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z\\"
                >
                </path>
              </svg>
            </button>
          </div>
        </div>
        <div id=\\"react-autowhatever-1\\"
             role=\\"listbox\\"
             class=\\"dropdown\\"
        >
        </div>
      </div>
    </form>
  </nav>
</div>
"
`;

exports[`<NavBar /> matches snapshot with 5 alerts 1`] = `
"
<div class=\\"container visible\\">
  <nav class=\\"navbar fixed-top navbar-expand navbar-dark p-1 bg-primary-transparent d-inline-block\\">
    <div style=\\"position:absolute;width:0;height:0;visibility:hidden;display:none\\">
    </div>
    <span class=\\"navbar-brand my-0 mx-2 h1 d-none d-sm-block float-left\\">
      5
      <svg aria-hidden=\\"true\\"
           data-prefix=\\"fas\\"
           data-icon=\\"circle-notch\\"
           class=\\"svg-inline--fa fa-circle-notch fa-w-16 fa-lg mx-1 text-muted\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 512 512\\"
           style=\\"opacity:0\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z\\"
        >
        </path>
      </svg>
    </span>
    <ul class=\\"navbar-nav float-right d-flex flex-row\\">
      <li class=\\"nav-item\\">
        <span class=\\"nav-link cursor-pointer\\">
          <svg aria-hidden=\\"true\\"
               data-prefix=\\"fas\\"
               data-icon=\\"bell-slash\\"
               class=\\"svg-inline--fa fa-bell-slash fa-w-20 \\"
               role=\\"img\\"
               xmlns=\\"http://www.w3.org/2000/svg\\"
               viewbox=\\"0 0 640 512\\"
          >
            <path fill=\\"currentColor\\"
                  d=\\"M633.82 458.1l-90.62-70.05c.19-1.38.8-2.66.8-4.06.05-7.55-2.61-15.27-8.61-21.71-19.32-20.76-55.47-51.99-55.47-154.29 0-77.7-54.48-139.9-127.94-155.16V32c0-17.67-14.32-32-31.98-32s-31.98 14.33-31.98 32v20.84c-40.33 8.38-74.66 31.07-97.59 62.57L45.47 3.37C38.49-2.05 28.43-.8 23.01 6.18L3.37 31.45C-2.05 38.42-.8 48.47 6.18 53.9l588.35 454.73c6.98 5.43 17.03 4.17 22.46-2.81l19.64-25.27c5.42-6.97 4.17-17.02-2.81-22.45zM157.23 251.54c-8.61 67.96-36.41 93.33-52.62 110.75-6 6.45-8.66 14.16-8.61 21.71.11 16.4 12.98 32 32.1 32h241.92L157.23 251.54zM320 512c35.32 0 63.97-28.65 63.97-64H256.03c0 35.35 28.65 64 63.97 64z\\"
            >
            </path>
          </svg>
        </span>
      </li>
      <li class=\\"nav-item\\">
        <span class=\\"nav-link cursor-pointer\\">
          <svg aria-hidden=\\"true\\"
               data-prefix=\\"fas\\"
               data-icon=\\"cog\\"
               class=\\"svg-inline--fa fa-cog fa-w-16 \\"
               role=\\"img\\"
               xmlns=\\"http://www.w3.org/2000/svg\\"
               viewbox=\\"0 0 512 512\\"
          >
            <path fill=\\"currentColor\\"
                  d=\\"M444.788 291.1l42.616 24.599c4.867 2.809 7.126 8.618 5.459 13.985-11.07 35.642-29.97 67.842-54.689 94.586a12.016 12.016 0 0 1-14.832 2.254l-42.584-24.595a191.577 191.577 0 0 1-60.759 35.13v49.182a12.01 12.01 0 0 1-9.377 11.718c-34.956 7.85-72.499 8.256-109.219.007-5.49-1.233-9.403-6.096-9.403-11.723v-49.184a191.555 191.555 0 0 1-60.759-35.13l-42.584 24.595a12.016 12.016 0 0 1-14.832-2.254c-24.718-26.744-43.619-58.944-54.689-94.586-1.667-5.366.592-11.175 5.459-13.985L67.212 291.1a193.48 193.48 0 0 1 0-70.199l-42.616-24.599c-4.867-2.809-7.126-8.618-5.459-13.985 11.07-35.642 29.97-67.842 54.689-94.586a12.016 12.016 0 0 1 14.832-2.254l42.584 24.595a191.577 191.577 0 0 1 60.759-35.13V25.759a12.01 12.01 0 0 1 9.377-11.718c34.956-7.85 72.499-8.256 109.219-.007 5.49 1.233 9.403 6.096 9.403 11.723v49.184a191.555 191.555 0 0 1 60.759 35.13l42.584-24.595a12.016 12.016 0 0 1 14.832 2.254c24.718 26.744 43.619 58.944 54.689 94.586 1.667 5.366-.592 11.175-5.459 13.985L444.788 220.9a193.485 193.485 0 0 1 0 70.2zM336 256c0-44.112-35.888-80-80-80s-80 35.888-80 80 35.888 80 80 80 80-35.888 80-80z\\"
            >
            </path>
          </svg>
        </span>
      </li>
    </ul>
    <form class=\\"form-inline mw-100\\"
          data-filters
    >
      <div role=\\"combobox\\"
           aria-haspopup=\\"listbox\\"
           aria-owns=\\"react-autowhatever-1\\"
           aria-expanded=\\"false\\"
           class=\\"autosuggest d-inline-block w-100\\"
      >
        <div class=\\"input-group mr-2\\">
          <div class=\\"input-group-prepend\\">
            <span class=\\"input-group-text px-2\\">
              <svg aria-hidden=\\"true\\"
                   data-prefix=\\"fas\\"
                   data-icon=\\"search\\"
                   class=\\"svg-inline--fa fa-search fa-w-16 \\"
                   role=\\"img\\"
                   xmlns=\\"http://www.w3.org/2000/svg\\"
                   viewbox=\\"0 0 512 512\\"
              >
                <path fill=\\"currentColor\\"
                      d=\\"M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z\\"
                >
                </path>
              </svg>
            </span>
          </div>
          <div class=\\"form-control p-1 components-filterinput\\">
            <input type=\\"text\\"
                   class=\\"components-filterinput-wrapper\\"
                   placeholder
                   value
                   autocomplete=\\"off\\"
                   aria-autocomplete=\\"list\\"
                   aria-controls=\\"react-autowhatever-1\\"
            >
          </div>
          <div class=\\"input-group-append\\">
            <button class=\\"input-group-text rounded-right cursor-pointer components-navbar-history px-2\\"
                    type=\\"button\\"
                    data-toggle=\\"dropdown\\"
                    aria-haspopup=\\"true\\"
                    aria-expanded=\\"true\\"
            >
              <svg aria-hidden=\\"true\\"
                   data-prefix=\\"fas\\"
                   data-icon=\\"caret-down\\"
                   class=\\"svg-inline--fa fa-caret-down fa-w-10 \\"
                   role=\\"img\\"
                   xmlns=\\"http://www.w3.org/2000/svg\\"
                   viewbox=\\"0 0 320 512\\"
              >
                <path fill=\\"currentColor\\"
                      d=\\"M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z\\"
                >
                </path>
              </svg>
            </button>
          </div>
        </div>
        <div id=\\"react-autowhatever-1\\"
             role=\\"listbox\\"
             class=\\"dropdown\\"
        >
        </div>
      </div>
    </form>
  </nav>
</div>
"
`;
